clang/lib/Sema/SemaType.cpp

```diff
@@ -9259,7 +9283,17 @@ static void HandleBoundsSafetyPointerTypeAttr(TypeProcessingState &state,
   if (auto AT = T->getAs<AtomicType>())
     T = AT->getValueType();
 
-  if (!T->isPointerType() && !T->getAs<AutoType>()) {
+  // In attribute-only mode we lift the restriction that `T` must be a pointer
```

Comment:
Why only in attribute-only mode?

Reply:
Because in that mode `__single` and `__unsafe_indexable` are sugar types. In `-fbounds-safety` that information is embedded in `PointerType` instead. Given that there's no good way to make this work in `-fbounds-safety` right now I decided to limit this to attribute-only mode. This seems like the right call given that it's the only one that really matters right now.

Reply:
I'm not sure we *need* to implement this for `-fbounds-safety` now, but I'm not sure that there isn't any good way to make that work.

Reply:
Ok there is a way. As you outlined elsewhere we could represent `__single` and `__unsafe_indexable` as sugar types in the AST for uninstantiated templates and then at template instantiation time rewrite the AST so that the sugar is dropped and the appropriate attribute is set in the `PointerType`.

This doesn't seem good to me because we'd be using two different representations for "is single" in the AST which might lead to bugs. Probably more importantly, I don't think this is an important problem to solve right now.